Rewrite math jit helpers to managed code

MichalPetryka opened the pull request:
Rewrites math helpers that has multiple implementations (C++ for CoreCLR, Windows X86 asm for CoreCLR and C#/C++ for NativeAOT) into a single, shared C# implementation.
I've put the new helpers in `Math` and `MathF` but some other place like `RuntimeHelpers` could perhaps be better.

On 64bit platforms this should be mostly the same, some direct helper calls might be replaced with indirect ones when the helpers didn't get to T1 yet.

Might result in slight perf regressions on X86 and ARM32 due to removal of asm paths and the managed helpers needing to do another call in some cases.

Regressions could be made up by converting casts to calls earlier and making them inlineable, that's not handled here to keep the changes simple.

This removes most of the implementation of 4 unused helpers, I think I forwarded them to temporary managed variants for legacy R2R images correctly but verification would be appreciated.

MichalPetryka committed:
Rewrite math jit helpers to managed code

MichalPetryka committed:
Fix asserts

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/Math.cs:
```diff
@@ -1481,5 +1481,251 @@ public static double ScaleB(double x, int n)
+            return 0;
+        }
+
+        private static double ULongToDouble(ulong val)
+        {
+            double conv = (long)val;
+            if (conv < 0)
+                conv += 4294967296.0 * 4294967296.0;  // add 2^64
+            Debug.Assert(conv >= 0);
+            return conv;
+        }
```
Interesting note: I experimented uint64->double cast with integer bit operations. It's about 3x faster than FP ALU on my machine.

MichalPetryka replied:
This doesn't use the FPU on X64 today, only SSE2.

huoyaoyuan replied:
I mean 3x faster than this implementation. But I don't expect it to be a consistent result around different CPUs.

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/Math.cs:
```diff
@@ -1481,5 +1481,251 @@ public static double ScaleB(double x, int n)
+            return FMod(dividend, divisor);
+        }
+
+        private static int DoubleToInt(double val) => (int)(long)val;
+        private static uint DoubleToUInt(double val) => (uint)(long)val;
```
Do we really need helpers this simple? Is it unsuitable for JIT to expand one cast into two?

MichalPetryka replied:
Those are needed for existing R2R binaries and can be removed after a major R2R version bump.

MichalPetryka commented:
@MihuBot

jkotas commented:
> This removes most of the implementation of 4 unused helpers, I think I forwarded them to temporary managed variants for legacy R2R images correctly but verification would be appreciated.

It would be better to leave these no longer used helpers alone, and delete them next time we rev min R2R version.

MichalPetryka committed:
Merge branch 'main' into math-helpers

MichalPetryka committed:
Merge branch 'main' into math-helpers

MichalPetryka committed:
Merge remote-tracking branch 'upstream/main' into math-helpers

khushal1996 commented:
@MichalPetryka 
Just wanted to ask if there is a plan to merge this PR soon. Since https://github.com/dotnet/runtime/pull/97529 will depend on this PR. Any idea when the current PR will be merged?

MichalPetryka committed:
Cleanup changes, add comments

MichalPetryka commented:
> @MichalPetryka Just wanted to ask if there is a plan to merge this PR soon. Since #97529 will depend on this PR. Any idea when the current PR will be merged?

I've cleaned up the changes now and added some comments, I've also verified that .NET 8 X86 R2R binaries seem to still be working so I think this is only waiting for review now?
One question would be whether the helpers should be moved to some other file.
cc @tannergooding @jkotas

MichalPetryka committed:
Fix typo

BruceForstall commented on src/coreclr/inc/jithelpers.h:
```diff
@@ -47,26 +45,26 @@
     JITHELPER(CORINFO_HELP_LMUL,                JIT_LMul,           CORINFO_HELP_SIG_16_STACK)
-    JITHELPER(CORINFO_HELP_LMUL_OVF,            JIT_LMulOvf,        CORINFO_HELP_SIG_16_STACK)
-    JITHELPER(CORINFO_HELP_ULMUL_OVF,           JIT_ULMulOvf,       CORINFO_HELP_SIG_16_STACK)
+    DYNAMICJITHELPER(CORINFO_HELP_LMUL_OVF,     NULL,               CORINFO_HELP_SIG_16_STACK)
```
You need to change the JIT-EE GUID

MichalPetryka committed:
Fix small type overflow in the scanner

MichalPetryka committed:
Update JIT-EE GUID

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/IL/HelperExtensions.cs:
```diff
@@ -40,6 +40,13 @@ public static MethodDesc GetOptionalHelperEntryPoint(this TypeSystemContext cont
             return helperMethod;
         }
 
+        public static MethodDesc GetHelperEntryPoint(this TypeSystemContext context, string typeNamespace, string typeName, string methodName)
```
Not a fan of this. This is more like "get arbitrary thing in corelib".

I see we already have to do `context.SystemModule.GetKnownType("System", "Math")` in the only place that uses this anyway. I'd just create a static local function to find a method on `System.Math` there and use that.

MichalPetryka replied:
Removed.

MichalPetryka committed:
Remove helper

jkotas commented:
> Might result in slight perf regressions on X86

I have done a quick adhoc perf test for ulong -> double conversions on x86:

```csharp
using System;
using System.Diagnostics;

for (;;)
{
    Stopwatch sw = new();
    sw.Start();
    Work(10_000_000_000);
    Internal.Console.WriteLine(sw.ElapsedMilliseconds.ToString());
}

static double Work(ulong iters)
{
    double sum = 0;
    for (ulong i = 0; i < iters; i += 42)
    {
        sum += i;
    }
    return sum;
}
``` 

I am seeing 1.5+x perf regression on x86 with this change (~640ms per iteration before this change, 1000ms per iteration with this change).

As you know, I am very supportive of rewriting stuff in C# but I do not think we can take perf regressions of this magnitude for the rewrite.

jkotas commented:
Perf for some of the rewritten operations is fine. Could you please go over each operation and keep only the ones that are not regressing perf significantly (let's say 10%)?

MichalPetryka committed:
Merge remote-tracking branch 'upstream/main' into math-helpers

MichalPetryka commented:
I've tried to remove the ulong -> double helper by expanding the conversion in the JIT but I've been unable to to do in a simple way, I'll restore the native helper tomorrow.
As for the other helpers converting from double, those will be reworked with the changes that make them do saturating conversion made by the Intel team so I wouldn't matter about them for now.
With the `rem` helpers, I'll need to look into them tomorrow but I assume the perf difference won't be too big and @tannergooding earlier mentioned that we could implement FMod in managed which would mitigate them.

MichalPetryka commented:
> I am seeing 1.5+x perf regression on x86 with this change (~640ms per iteration before this change, 1000ms per iteration with this change).

Does your CPU have AVX512? I've looked into what the MSVC helper does and it apparently cheats with that (which makes things annoying to benchmark since my CPU has it).

jkotas commented:
No AVX512. Intel(R) Core(TM) i9-10900K

MichalPetryka committed:
Fully managed conversions to double

MichalPetryka commented:
> > Might result in slight perf regressions on X86
> 
> I have done a quick adhoc perf test for ulong -> double conversions on x86:
> 
> ```cs
> using System;
> using System.Diagnostics;
> 
> for (;;)
> {
>     Stopwatch sw = new();
>     sw.Start();
>     Work(10_000_000_000);
>     Internal.Console.WriteLine(sw.ElapsedMilliseconds.ToString());
> }
> 
> static double Work(ulong iters)
> {
>     double sum = 0;
>     for (ulong i = 0; i < iters; i += 42)
>     {
>         sum += i;
>     }
>     return sum;
> }
> ```
> 
> I am seeing 1.5+x perf regression on x86 with this change (~640ms per iteration before this change, 1000ms per iteration with this change).
> 
> As you know, I am very supportive of rewriting stuff in C# but I do not think we can take perf regressions of this magnitude for the rewrite.

I've come up with a new, fully managed implementation for both long and ulong conversion that's slightly faster than current on 2 machines that I've tested on (It's noticeably faster when adapted for uint but we don't have a helper for that today). Could you verify you no longer see a regression on your machine @jkotas?

jkotas commented:
> Could you verify you no longer see a regression on your machine @jkotas?

Yes, it works better. 1.2x faster compared to baseline on my microbenchmark on my machine.

Does this implementation work well for Arm32 too or is it x86 specific?

MichalPetryka commented:
> Does this implementation work well for Arm32 too or is it x86 specific?

On Arm32 we have casts from `uint` so we don't need the adds here. Do you want to add an `#if` for platforms with casts from `uint`?

jkotas commented on src/coreclr/inc/corinfo.h:
```diff
@@ -387,18 +387,18 @@ enum CorInfoHelpFunc
     CORINFO_HELP_DBL2LNG,
     CORINFO_HELP_DBL2LNG_OVF,
-    CORINFO_HELP_DBL2UINT,
+    CORINFO_HELP_DBL2UINT,              // unused
```
I expect that #97529 is going to add dependency on this helper back.

MichalPetryka replied:
I think it makes more sense for them to remove the comment if this will be merged first and they'll end up needing the helper again.

jkotas replied:
My preference would be to leave all changes related to floating point to integer conversions to the Intel PRs.

jkotas commented on src/coreclr/inc/corinfo.h:
```diff
@@ -387,18 +387,18 @@ enum CorInfoHelpFunc
     CORINFO_HELP_LNG2DBL,               // Convert a signed int64 to a double
     CORINFO_HELP_ULNG2DBL,              // Convert a unsigned int64 to a double
-    CORINFO_HELP_DBL2INT,
+    CORINFO_HELP_DBL2INT,               // unused
```
```suggestion
    CORINFO_HELP_DBL2INT,
```
Reduce conflicts with https://github.com/dotnet/runtime/pull/97529

filipnavara commented:
> Does this implementation work well for Arm32 too or is it x86 specific?

I'll run some benchmarks on ARM hardware once I get back to office.

jkotas commented on src/coreclr/vm/i386/jitinterfacex86.cpp:
```diff
@@ -961,32 +942,6 @@ void InitJITHelpers1()
-        SetJitHelperFunction(CORINFO_HELP_DBL2INT, JIT_Dbl2IntSSE2);
-        if (dwCPUFeaturesECX & 1)  // check SSE3
-        {
-            SetJitHelperFunction(CORINFO_HELP_DBL2UINT, JIT_Dbl2LngSSE3);
```
Dbg2Long conversion have 1.6+x perf regression on my machine:
```
using System;
using System.Diagnostics;

for (;;)
{
    Stopwatch sw = new();
    sw.Start();
    Work(10_000_000_000);
    Console.WriteLine(sw.ElapsedMilliseconds.ToString());
}

static long Work(double iters)
{
    long sum = 0;
    for (double i = 0; i < iters; i += 42)
    {
        sum += (long)i;
    }
    return sum;
}
```

MichalPetryka replied:
I don't think this matters much since Intel people want to remove the asm implemementation anyway for the change to saturating, with MSVC they'll probably make the C++ helper call `_cvt_dtoll_sat` I guess. I agree with @tannergooding that any performance comparisons should be done when they're finished.

jkotas replied:
Right, so it would be best to avoid changing and introducing regressions in the floating-point to integer conversions in this PR and leave all that for the Intel PR(s). Spreading changes like that into PRs with unrelated changes makes accounting more difficult.

MichalPetryka replied:
I've noticed something interesting, this seems to regress 25x on a CPU with AVX512, seems like there are some bad MSVC bugs here.

MichalPetryka replied:
I've restored the SSE3 FPU helper, SSE2 only CPUs should be so rare that it makes no sense to keep separate paths for them these days I'd say.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10722,12 +10733,15 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
                     {
                         // Cache it for future uses to avoid taking the lock again.
                         hlpFinalTierAddrTable[dynamicFtnNum] = finalTierAddr;
+                        EE_TO_JIT_TRANSITION_LEAF();
```
This should use `goto exit` pattern used elsewhere and/or factored into a helper method. These macros are not meant to be used multiple times per method like this. It happens to work for the LEAF macro, but it would not work for the non-LEAF macro.

jkotas replied:
There is one more return earlier in the method that should get the same treatment.

MichalPetryka committed:
Merge remote-tracking branch 'upstream/main' into math-helpers

MichalPetryka committed:
Restore one ASM helper

jkotas commented on src/coreclr/jit/morph.cpp:
```diff
@@ -377,7 +377,9 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
                         return nullptr;
 #else  // TARGET_X86
-                        return fgMorphCastIntoHelper(tree, CORINFO_HELP_DBL2UINT, oper);
+                        oper = gtNewCastNode(TYP_LONG, oper, false, TYP_LONG);
```
It would be better to leave this change to the Intel PRs too.

MichalPetryka replied:
The previous behaviour here was kinda a bug since `CORINFO_HELP_DBL2UINT` was bound to `Dbl2Lng`, it only worked fine cause X86 and ARM32 ABIs return the lower bits of the long same as an uint. Calling the `Dbl2UInt` helper here would be a performance regression which you said are better left to the Intel PRs.

jkotas replied:
Yes, I agree the existing code is a hack but there is no observable bug. There is no point in cleaning it up in this PR. It requires an extra throw away work to evaluate the impact. I suspect that this change may be changing the observable behavior. We want to have the changes of observable behavior centralized in the Intel PR.

This PR is 95% good and non-controversial, and it is just these few changes that are holding it back. If you would like it to merged soon, I would recommend reverting the few controversial changes that are holding it back.

MichalPetryka replied:
> I suspect that this change may be changing the observable behavior.

How so? Isn't casting a long to uint the same as taking the lower 4B from it?

> these few changes

You mean the change here and the `// unused` comments?

MichalPetryka replied:
> Ideally, anything related to floating point to integer conversions that are being completely redone in the Intel PR.

I'd argue that cleaning things up here makes the Intel changes simpler since it reduces the amount of work they need to do there. If the Intel folks prefer it, I can remove those from here however cc @khushal1996

To avoid JIT changes in this PR, I've moved `double -> uint` casts to a managed implementation in the latest [that copies what LLVM does](https://godbolt.org/z/vKerYs1a8) since it's faster on my machine anyway. It is a bit of a throwaway work like you said but this change makes the PR simpler and I'd argue that it's safe since LLVM uses this implementation.

filipnavara commented:
Running the benchmark from https://github.com/dotnet/runtime/pull/98858#issuecomment-1986379186 on Raspberry Pi 5 ARM32 - Baseline: ~4410, PR: ~3116.

Any other operation that should be tested?

MichalPetryka commented:
> Running the benchmark from [#98858 (comment)](https://github.com/dotnet/runtime/pull/98858#issuecomment-1986379186) on Raspberry Pi 5 ARM32 CoreCLR - Baseline: ~4410, PR: ~3116.

@jkotas Do we want to keep it as is then or do we still want to optimize it with `#if` for ARM32?

Also, do we want to keep the managed helpers in Math in the end or do we want to move them somewhere?

jkotas commented:
> @jkotas Do we want to keep it as is then or do we still want to optimize it with #if for ARM32?

What would be the optimization for Arm32? The PR is faster than baseline already based on the @filipnavara measurement.

> Also, do we want to keep the managed helpers in Math in the end or do we want to move them somewhere?

I do not have an opinion. Math is fine with me.

jkotas commented:
ok, if it makes it even faster - sounds good to me. I would make it `#if TARGET_X86`. x86 tends to be the odd platform that requires special casing.

MichalPetryka commented:
> I would make it `#if TARGET_X86`. x86 tends to be the odd platform that requires special casing.

I wanted to do it like that at first too, going that way would however require all other architectures that'd use the helper to implement the uint casts in the JIT so this seems safer.

jkotas commented:
If there is an architecture that is like x86, we are going to find out fairly quickly (it won't work) and the ifdef can be adjusted for it. On the other hand, the other direction is hard to find (one has to be lucky to spot that helper is slower than it needs to be).

MichalPetryka committed:
Merge remote-tracking branch 'upstream/main' into math-helpers

MichalPetryka committed:
Cleanup code, move uint cast to managed

MichalPetryka committed:
Update jithelpers.h

MichalPetryka committed:
Remove DoubleToUInt

MichalPetryka committed:
Move the helpers

MichalPetryka commented:
I've moved the helpers according to the suggestions from @tannergooding made on discord, they now mirror the APIs approved in #61885.

MichalPetryka committed:
Remove the using

jkotas commented on src/coreclr/inc/jiteeversionguid.h:
```diff
@@ -43,11 +43,11 @@ typedef const GUID *LPCGUID;
-    0xedc8,
-    0x48f1,
-    {0xa7, 0x5d, 0x0f, 0xa1, 0xf0, 0xb5, 0xaf, 0x6f}
+constexpr GUID JITEEVersionIdentifier = { /* f4da6024-dd94-4b02-af3d-83e74297b17b */
```
Reving JIT/EE GUID should not be necessary anymore.

MichalPetryka replied:
Are you sure? This PR still changed `JITHELPER`s into `DYNAMICJITHELPER`s which AFAIR made @BruceForstall say it's needed before.

MichalPetryka commented on src/coreclr/vm/corelib.h:
```diff
@@ -259,11 +259,36 @@ DEFINE_FIELD(DELEGATE,            METHOD_PTR_AUX,         _methodPtrAux)
 
-DEFINE_CLASS(INT128,               System,                 Int128)
-DEFINE_CLASS(UINT128,              System,                 UInt128)
+DEFINE_CLASS(DOUBLE_SPC,            System,                 Double)
```
I had to add a suffix for the types here cause otherwise they caused build errors due to name conflicts. I used `_SPC` for `System.Private.CoreLib` but maybe there's something better here?

MichalPetryka committed:
Fix typos

jkotas commented on src/coreclr/vm/corelib.h:
```diff
@@ -259,11 +259,36 @@ DEFINE_FIELD(DELEGATE,            METHOD_PTR_AUX,         _methodPtrAux)
 
-DEFINE_CLASS(INT128,               System,                 Int128)
-DEFINE_CLASS(UINT128,              System,                 UInt128)
+DEFINE_CLASS(DOUBLE_SPC,            System,                 Double)
+DEFINE_METHOD(DOUBLE_SPC,           CONVERT_TO_INT32,       ConvertToInt32,             NoSig)
+DEFINE_METHOD(DOUBLE_SPC,           CONVERT_TO_UINT64,      ConvertToUInt64,            NoSig)
```
```suggestion
// Defined as a primitive type, avoid redefinition
// DEFINE_CLASS(DOUBLE,            System,                 Double)
DEFINE_METHOD(DOUBLE,           CONVERT_TO_INT32,       ConvertToInt32,             NoSig)
DEFINE_METHOD(DOUBLE,           CONVERT_TO_UINT64,      ConvertToUInt64,            NoSig)
...
```
Would this work?

MichalPetryka replied:
Doing this fails to build:
```cs
H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\CreateRuntimeRootILLinkDescriptorFile.targets(41,5): error MSB4018: Nastąpiło nieoczekiwane niepowodzenie zadania „CreateRuntimeRootILLinkDescriptorFile”. [H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\CreateRuntimeRootILLinkDescriptorFile.targets(41,5): error MSB4018: System.Collections.Generic.KeyNotFoundException->Microsoft.Build.Framework.BuildException.GenericBuildTransferredException: The given key 'DOUBLE' was not present in the dictionary. [H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\CreateRuntimeRootILLinkDescriptorFile.targets(41,5): error MSB4018:    at System.Collections.Generic.Dictionary`2.get_Item(TKey key) [H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\CreateRuntimeRootILLinkDescriptorFile.targets(41,5): error MSB4018:    at ILLink.Tasks.CreateRuntimeRootILLinkDescriptorFile.AddMethod(String methodName, String classId, String classNamespace, String className, Nullable`1 featureSwitch) in H:\Projects\dotnet\runtime\src\tools\illink\src\ILLink.Tasks\CreateRuntimeRootDescriptorFile.cs:line 438 [H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\CreateRuntimeRootILLinkDescriptorFile.targets(41,5): error MSB4018:    at ILLink.Tasks.CreateRuntimeRootILLinkDescriptorFile.ProcessMscorlib(String typeFile) in H:\Projects\dotnet\runtime\src\tools\illink\src\ILLink.Tasks\CreateRuntimeRootDescriptorFile.cs:line 241 [H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\CreateRuntimeRootILLinkDescriptorFile.targets(41,5): error MSB4018:    at ILLink.Tasks.CreateRuntimeRootILLinkDescriptorFile.Execute() in H:\Projects\dotnet\runtime\src\tools\illink\src\ILLink.Tasks\CreateRuntimeRootDescriptorFile.cs:line 136 [H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\CreateRuntimeRootILLinkDescriptorFile.targets(41,5): error MSB4018:    at Microsoft.Build.CommandLine.OutOfProcTaskAppDomainWrapperBase.InstantiateAndExecuteTask(IBuildEngine oopTaskHostNode, LoadedType taskType, String taskName, String taskLocation, String taskFile, Int32 taskLine, Int32 taskColumn, IDictionary`2 taskParams) [H:\Projects\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
```

jkotas replied:
The following should fix this build break:
```
#if FOR_ILLINK
DEFINE_CLASS(DOUBLE,            System,                 Double)
#endif
DEFINE_METHOD(DOUBLE,           CONVERT_TO_INT32,       ConvertToInt32,             NoSig)
...
```

jkotas commented on src/coreclr/vm/i386/jithelp.S:
```diff
@@ -552,55 +552,16 @@ LOCAL_LABEL(LRszMORE32):
-//    equal or later than Core2 and Atom that have SSE3 support
+//    This code is faster than C++ even on modern CPUs but it requires SSE3
 //
-LEAF_ENTRY JIT_Dbl2LngSSE3, _TEXT
```
What is `Fpu` suffix trying to say?

I think `JIT_Dbl2LngSSE3` was a better name for this method.

MichalPetryka replied:
Wanted to note it as using FPU instructions which are generally discouraged these days. I honestly don't know which one is the better name here.

jkotas replied:
I think that the SSE3 is a better name. The distinct characteristic of this implementation is that it requires SSE3. It is the best implementation possible assuming SSE3 as a baseline. I do not think that the details that it still uses x87 floating point instructions should be captured in the name.

(Consider hypothetical case where we would have multiple implementations of this helper for multiple different x86 baselines - AVX, AVX2, ... . Including the baseline ID in the name would be the most sensible naming scheme.)

jkotas replied:
Also, touching this helper is a throw away change. As I have said before, my preference would be to omit all throw-away changes from this PR.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Double.cs:
```diff
@@ -440,6 +440,162 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
+            return 0;
+        }
+
+        private static ref double DoubleConversionTable => ref MemoryMarshal.GetReference([0.0, UIntMaxValueOffset]);
```
Nit: Does this need to be factored into a helper method? Can it be inline?

jkotas replied:
Ah ok, you are using it in two places.

MichalPetryka replied:
Two methods use this so I wanted to avoid duplicating it in both.

MichalPetryka committed:
Rename defined and helper

MichalPetryka committed:
Merge remote-tracking branch 'upstream/main' into math-helpers

MichalPetryka commented:
@MihuBot

MichalPetryka committed:
Update jitinterface.h

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Double.cs:
```diff
@@ -440,6 +440,162 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
+        }
+
+        [StackTraceHidden]
+        private static ulong ConvertToUInt64Overflow(double val)
```
If we are aligning with the public surface naming conventions, these methods should have `Checked` suffix (prior art: `TryConvertFromChecked`,  `CreateChecked`, ...)

MichalPetryka replied:
Done.

MichalPetryka committed:
Rename to Checked

jkotas commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -5973,6 +5973,7 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
                 switch (tree->AsIntrinsic()->gtIntrinsicName)
                 {
                     case NI_System_Math_Atan2:
+                    case NI_System_Math_FMod:
```
@dotnet/jit-contrib Could you please sign-off on this change?

EgorBo replied:
this change is safe, it's just slightly increases node's cost. I presume this whole part should be rewritten to use `IsIntrinsicImplementedByUserCall`

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:
```diff
@@ -340,32 +127,31 @@ public static long UMod(uint i, uint j)
         // of the hot path because of it does not need to raise full stackframe.
         //
-
         [MethodImpl(MethodImplOptions.NoInlining)]
```
If you are touching these helpers, you can also delete `[MethodImpl(MethodImplOptions.NoInlining)]` and change them to return `void`.

These are old workarounds from the times when the JIT was not smart enough about throw helpers.

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/JitHelper.cs:
```diff
@@ -164,54 +164,67 @@ public static void GetEntryPoint(TypeSystemContext context, ReadyToRunHelper id,
+                    mangledName = "RhpDbl2Lng";
+                    break;
+                case ReadyToRunHelper.Dbl2UInt:
+                    mangledName = "RhpDbl2Lng";
```
Replicating the CoreCLR hack here is not an improvement.

jkotas reviewed (approved):
LGTM modulo feedback. Thank you!

jkotas commented:
@tannergooding Could you please review this change as well?

MichalPetryka committed:
Cleanup NativeAOT, fix parameter names

MichalPetryka committed:
Update jitinterface.cpp

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:
```diff
@@ -19,147 +18,141 @@ internal static partial class MathHelpers
         }
 #endif // TARGET_ARM
 
         //
         // Matching return types of throw helpers enables tailcalling them. It improves performance
         // of the hot path because of it does not need to raise full stackframe.
         //
```
```suggestion
```

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:
```diff
@@ -19,147 +18,141 @@ internal static partial class MathHelpers
         //
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static long ThrowLngOvf()
+        private static void ThrowLngOvf()
```
You can actually delete these and replace them with calls to `ThrowHelper.ThrowDivideByZeroException`, `ThrowHelper.ThrowOverflowException`, .... .

MichalPetryka replied:
Ah yeah, missed this.

MichalPetryka committed:
Cleanup throw helpers

tannergooding commented:
> Could you please review this change as well?

Will do. It's first on my list to get to tomorrow and then I'll merge, assuming no other feedback

jkotas reviewed (changes requested):
The tests are failing:
```
Assert failure(PID 9608 [0x00002588], Thread: 9612 [0x258c]): !"Precode::GetPrecodeFromEntryPoint: Unexpected code in precode"

CORECLR! Precode::GetPrecodeFromEntryPoint + 0x4A (0x718b1743)
CORECLR! CEEJitInfo::getHelperFtn + 0x1EB (0x719160db)
```

MichalPetryka commented:
> The tests are failing:
> 
> ```
> Assert failure(PID 9608 [0x00002588], Thread: 9612 [0x258c]): !"Precode::GetPrecodeFromEntryPoint: Unexpected code in precode"
> 
> CORECLR! Precode::GetPrecodeFromEntryPoint + 0x4A (0x718b1743)
> CORECLR! CEEJitInfo::getHelperFtn + 0x1EB (0x719160db)
> ```

Ah I might know what this is, it seems to not like the `Dbl2UInt` helper being unmanaged in CoreCLR after I ported it to managed in NativeAOT again.

MichalPetryka committed:
Remove the unmanaged helper from the tiering optimization

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Double.cs:
```diff
@@ -440,6 +440,163 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
+
+        // TODO: make the helpers saturating when implementing https://github.com/dotnet/runtime/issues/61885
+        private static int ConvertToInt32(double value) => (int)(long)value;
+        private static uint ConvertToUInt32(double value) => (uint)(long)value;
```
What is the performance of this implementation compared to RhpDbl2Int/RhpDbl2UInt that was used before this change?

MichalPetryka replied:
`RhpDbl2Int` is unused today AFAIK and `RhpDbl2UInt` will only be used by X86 which @filipnavara is currently exposing so there'd be no observable regression.
As such, I'd rather keep things simple here and make it slightly slower for NAOT X86 due to the additional call and let it get fixed up with either helper inlining or the Intel saturating changes. Speaking of those, changing it to a managed implementation in NAOT makes it so that they don't have to test changes to a C++ helper that's only used on a non fully supported platform.

jkotas replied:
I would like to see the impact of these changes to be measured, or the changes to be reverted.

filipnavara replied:
Baseline: 573
PR: 630

for this code:
```csharp
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

for (;;)
{
    Stopwatch sw = new();
    sw.Start();
    Work(10_000_000_000);
    Console.WriteLine(sw.ElapsedMilliseconds.ToString());
}

[MethodImpl(MethodImplOptions.NoInlining)]
static uint Work(double iters)
{
    uint sum = 0;
    for (double i = 0; i < iters; i += 42)
    {
        sum += (uint)i;
    }
    return sum;
}
```

jkotas replied:
I do not see a good reason for taking this regression.

filipnavara commented:
We are missing this in win-x86 NativeAOT:
```
--- a/src/coreclr/nativeaot/Runtime/MathHelpers.cpp
+++ b/src/coreclr/nativeaot/Runtime/MathHelpers.cpp
@@ -271,6 +271,14 @@ FCIMPL1_F(float, tanhf, float x)
     return std::tanhf(x);
 FCIMPLEND

+FCIMPL2_DD(double, fmod, double x, double y)
+    return std::fmod(x, y);
+FCIMPLEND
+
+FCIMPL2_FF(float, fmodf, float x, float y)
+    return std::fmodf(x, y);
+FCIMPLEND
+
 FCIMPL3_DDD(double, fma, double x, double y, double z)
     return std::fma(x, y, z);
 FCIMPLEND
```

Also, the throw helpers are not properly rooted in the scanning phase:
```
    *** Methods compiled but not scanned:
    int64.MultiplyChecked(int64,int64)
    [S.P.CoreLib]System.ThrowHelper.ThrowDivideByZeroException()
    uint64.MultiplyChecked(uint64,uint64)
    [S.P.CoreLib]System.Text.StringBuilder+AppendInterpolatedStringHandler.AppendFormatted<uint64>(uint64)
    [S.P.CoreLib]System.Text.StringBuilder+AppendInterpolatedStringHandler.AppendFormattedWithTempSpace<uint64>(uint64,in
  t32,string)
    [S.P.CoreLib]System.Text.StringBuilder+AppendInterpolatedStringHandler.AppendCustomFormatter<uint64>(uint64,string)
    Unhandled exception: System.Exception: Scanning failure
       at ILCompiler.Program.Run() in D:\runtime\src\coreclr\tools\aot\ILCompiler\Program.cs:line 651
       at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass236_0.<.ctor>b__0(ParseResult result) in D:\runtime\src\corec
  lr\tools\aot\ILCompiler\ILCompilerRootCommand.cs:line 292
       at System.CommandLine.Invocation.InvocationPipeline.Invoke(ParseResult parseResult)
```

MichalPetryka committed:
Go back to C++ helper for doubles, root things with scanner, add fmod

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:
```diff
@@ -1224,41 +1224,75 @@ private void ImportBinaryOperation(ILOpcode opcode)
-                        _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.LMulOfv), "_lmulovf");
-                        _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.ULMulOvf), "_ulmulovf");
-                    }
+                    _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.LMulOfv), "_lmulovf");
```
Are these helpers actually used on all architectures?

jkotas replied:
The condition around these helpers in the JIT is !TARGET_64BIT, so this check should match that.

https://github.com/dotnet/runtime/blob/ee501fb2c6ac901b761131b8c4760f74b5c18a62/src/coreclr/jit/morph.cpp#L8463-L8498

MichalPetryka replied:
I don't think so but they seemed to fail on x86 and I assume that in total they should have minimal size impact and it makes no sense to complicate the scanner with architecture checks for this single place since doing that requires checking each helper in the JIT when adding new platforms and such.

jkotas replied:
We have the scanner so that these things can be tailored.

If you were to go with the premise that these things have minimal size, we would root all helpers all the time and delete most the code here.

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/JitHelper.cs:
```diff
@@ -164,54 +164,67 @@ public static void GetEntryPoint(TypeSystemContext context, ReadyToRunHelper id,
+                        new MethodSignature(MethodSignatureFlags.Static, 0, floatType, [floatType]));
+                    break;
+
+                case ReadyToRunHelper.Dbl2UInt:
+                    mangledName = "RhpDbl2ULng";
+                    break;
```
```suggestion
                case ReadyToRunHelper.Dbl2Int:
                    mangledName = "RhpDbl2Int";
                    break;
                case ReadyToRunHelper.Dbl2UInt:
                    mangledName = "RhpDbl2UInt";
                    break;
```

MichalPetryka committed:
Update MathHelpers.cpp

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/JitHelper.cs:
```diff
@@ -164,54 +164,67 @@ public static void GetEntryPoint(TypeSystemContext context, ReadyToRunHelper id,
-                case ReadyToRunHelper.Dbl2ULng:
-                    mangledName = "RhpDbl2ULng";
-                    break;
                 case ReadyToRunHelper.Dbl2Int:
```
```suggestion
```

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/JitHelper.cs:
```diff
@@ -164,54 +164,67 @@ public static void GetEntryPoint(TypeSystemContext context, ReadyToRunHelper id,
-                    break;
                 case ReadyToRunHelper.Dbl2Int:
-                    mangledName = "RhpDbl2Int";
+                    methodDesc = context.SystemModule.GetKnownType("System", "Double").GetKnownMethod("ConvertToInt32", null);
```
```suggestion
```

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/JitHelper.cs:
```diff
@@ -164,54 +164,67 @@ public static void GetEntryPoint(TypeSystemContext context, ReadyToRunHelper id,
                 case ReadyToRunHelper.Dbl2Int:
-                    mangledName = "RhpDbl2Int";
+                    methodDesc = context.SystemModule.GetKnownType("System", "Double").GetKnownMethod("ConvertToInt32", null);
                     break;
```
```suggestion
```

jkotas commented on src/coreclr/vm/corelib.h:
```diff
@@ -259,11 +259,44 @@ DEFINE_FIELD(DELEGATE,            METHOD_PTR_AUX,         _methodPtrAux)
+#ifdef FOR_ILLINK
+DEFINE_CLASS(DOUBLE,                System,                 Double)
+#endif
+DEFINE_METHOD(DOUBLE,               CONVERT_TO_INT32,       ConvertToInt32,             NoSig)
```
```suggestion
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Double.cs:
```diff
@@ -440,6 +440,162 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
+        private const double IntMaxValueOffset = 2147483648.0; // 2^31, int.MaxValue + 1
+        private const double UIntMaxValueOffset = 4294967296.0; // 2^32, uint.MaxValue + 1
+
+        // TODO: make the helpers saturating when implementing https://github.com/dotnet/runtime/issues/61885
+        private static int ConvertToInt32(double value) => (int)(long)value;
+
```
```suggestion
```

jkotas commented on src/coreclr/nativeaot/Runtime/MathHelpers.cpp:
```diff
@@ -106,33 +38,14 @@ EXTERN_C uint64_t QCALLTYPE RhpULMod(uint64_t i, uint64_t j)
-}
-FCIMPLEND
-
 FCIMPL1_D(uint32_t, RhpDbl2UInt, double val)
```
```suggestion
FCIMPL1_D(int32_t, RhpDbl2Int, double val)
{
    return (int32_t)val;
}
FCIMPLEND

FCIMPL1_D(uint32_t, RhpDbl2UInt, double val)
```

MichalPetryka replied:
CoreCLR code needs to be changed too if you want to go back to C++ since it was also switched to the C# variant for this unused helper.

jkotas replied:
Yes, that's fine

MichalPetryka committed:
Move unused helper back to C++, restore scanner ifs

MichalPetryka committed:
Update ILImporter.Scanner.cs

tannergooding commented on src/coreclr/inc/jithelpers.h:
```diff
@@ -47,26 +45,29 @@
-    JITHELPER(CORINFO_HELP_DBL2INT_OVF,         JIT_Dbl2IntOvf,     CORINFO_HELP_SIG_8_STACK)
+    DYNAMICJITHELPER(CORINFO_HELP_LNG2DBL,      NULL,               CORINFO_HELP_SIG_8_STACK)
+    DYNAMICJITHELPER(CORINFO_HELP_ULNG2DBL,     NULL,               CORINFO_HELP_SIG_8_STACK)
+    JITHELPER(CORINFO_HELP_DBL2INT,             JIT_Dbl2Int,        CORINFO_HELP_SIG_8_STACK)
```
What platform is this needed for?

Arm, Arm64, x86, and x64 all provide a direct instruction for `double` to `int32` as part of the baseline ISAs we require.

MichalPetryka replied:
It's unused but can't be removed without an R2R bump. Also, Intel PRs might need it again.

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:
```diff
@@ -1259,6 +1259,53 @@ private void ImportBinaryOperation(ILOpcode opcode)
                     {
                         _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.ThrowDivZero), "_divbyzero");
```
```suggestion
                        _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.ThrowDivZero), "_divbyzero");
                        if (opcode == ILOpcode.rem)
                            _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.Overflow), "_ovf");
```
Preexisting bug. Also, the same is missing for div above;

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:
```diff
@@ -1259,6 +1259,53 @@ private void ImportBinaryOperation(ILOpcode opcode)
+                    _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.Lng2Dbl), "conv_r");
+                    _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.ULng2Dbl), "conv_r");
                     break;
             }
```
```suggestion
            }
            if (checkOverflow)
            {
                _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.Overflow), "_ovf");
            }
```
Looks like pre-existing bug

MichalPetryka committed:
Update ILImporter.Scanner.cs

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:
```diff
@@ -1245,6 +1245,10 @@ private void ImportBinaryOperation(ILOpcode opcode)
                     {
                         _dependencies.Add(GetHelperEntrypoint(ReadyToRunHelper.ThrowDivZero), "_divbyzero");
                     }
+                    if (opcode == ILOpcode.rem)
```
```suggestion
                    if (opcode == ILOpcode.div)
```

jkotas replied:
Also, it should be under ARM64-only condition.

MichalPetryka committed:
Update src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Single.cs:
```diff
@@ -435,6 +435,33 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
+            //   the result is [dividend] (negated for -infinity***).
+            //
+            // ***"negated for -infinity" has been removed from the spec
+            if (divisor == 0 || !IsFinite(dividend))
+            {
+                return NaN;
+            }
+
+            if (!IsFinite(divisor) && !IsNaN(divisor))
+            {
+                return dividend;
+            }
```
Is this needed? C requires (for IEC 60559; aka IEEE 754; compatible implementations, which all our dependencies have to be):
* If x is ±0 and y is not zero, ±0 is returned.
* If x is ±∞ and y is not NaN, NaN is returned and FE_INVALID is raised.
* If y is ±0 and x is not NaN, NaN is returned and FE_INVALID is raised.
* If y is ±∞ and x is finite, x is returned.
* If either argument is NaN, NaN is returned.

So it looks like we can just defer to `FMod` directly and bypass this extra logic

MichalPetryka replied:
`If y is ±∞ and x is finite, x is returned.`
The `x is finite` part doesn't seem to be checked here so it'd different it seems.

tannergooding replied:
Values can be `finite`, `infinite`, or `nan`.

The implementation here is doing:
* If `divisor` (`y`) is zero, return nan
  * i.e. the third/fifth bullet points
* if `dividend` (`x`) is not finite (in other words, `x` is `nan` or `infinite`), return `NaN`
  * i.e. the second/fifth bullet points
* if `divisor` (`y`) is not `finite` and not `nan` (in other words, `y` is `infinite`), return `dividend` (`x`)
  * i.e, the 4th bullet point
  * `x` being `nan` or `infinite` is covered by the check I just described previously

Everything else defers down to `fmod`

MichalPetryka replied:
Ah I missed that it's `||` not `&&` there, gonna make it use FMod directly then.

MichalPetryka replied:
I've made the helper be `Math.FMod` directly in the last commit, in case it'll be rewritten in C# it should be then added back in `jitinterface.cpp` to have the JIT emit direct calls to it.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/UInt64.cs:
```diff
@@ -176,6 +177,58 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
+        //
+
+        [StackTraceHidden]
+        private static ulong MultiplyChecked(ulong left, ulong right)
```
This seems like a very expensive way to implement `MultiplyChecked`.

For `UInt128` we simply do `BigMul` and check if `upper` is 0: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/UInt128.cs,1379

But more practically speaking, we can determine overflow just by doing `if ((128 - LeadingZeroCount(left) - LeadingZeroCount(right)) > 64)`. Since `x-bit * y-bit` will produce an `(x + y)-bit` product; anything else must fit into the given 64-bits.

MichalPetryka replied:
I'd prefer to change logic here in a separate PR. Keep in mind that this is for 32bit platforms so things like BigMul are slow there.

tannergooding replied:
> I'd prefer to change logic here in a separate PR

That's fine.

> Keep in mind that this is for 32bit platforms so things like BigMul are slow there.

That's why I suggested the extremely cheap check of `if ((128 - LeadingZeroCount(left) - LeadingZeroCount(right)) > 64)`. That, on a 32-bit platform, ends up being an almost free check to determine overflow without doing any expensive multiplication (it simply uses the `lzcnt` equivalent that every platform we target, including WASM, natively supports).

Then, you only have to do a regular 64x64-bit multiplication since you know it can't overflow. You don't have to do the extra work (and many branches) to check overflow at each step in the algorithm.

jkotas replied:
> That, on a 32-bit platform, ends up being an almost free check to determine overflow without doing any expensive multiplication 

It looks like an approximate check to determine overflow. If it fails, you still need to take the more expensive multiplication to determine whether it is actually going to overflow - correct?

The current implementation has a variant of this idea where it checks whether the high 32-bits of both numbers are zero. If they are, it knows that there is not going to be an overflow and so it multiplies the low 32-bit portions and returns. If they are not, it does the more expensive multiplication.

I think the difference between what's there currently and your suggestion is in the speed of the fast path and the curoff for the fast path. The current implementation has lower cutoff and faster fast path. Your suggestion has higher cutoff and slower fast path.

tannergooding replied:
> It looks like an approximate check to determine overflow.

Yes.

> you still need to take the more expensive multiplication to determine whether it is actually going to overflow

Not quite.

--------------

In general, for an `x-bit * y-bit` value (excepting where one input is `0`), you need a minimum of `max(x, y)` bits to represent the product and a maximum of `x + y` bits.

The minimum bit count comes about because of the special case `1`, which returns the original value. Multiplication then adds one additional bit to the output for each power of 2 (multiplying by 2 is equivalent to shifting left by 1, multiplying by 4 is equivalent to shifting left by 2, thus multiplying by 3 must be between these two values and takes less than 2 additional bits).

Thus the exact number of bits needed is effectively `max(x, y) + ceil(log2(min(x, y)))` bits.

Finding the input with the smaller length and computing its effective `ceil(log2)` are both incredibly cheap operations that all our targets have direct acceleration to support. So it should always be significantly cheaper than what we're doing here already.

The approach I detailed of giving `if ((128 - LeadingZeroCount(left) - LeadingZeroCount(right)) > 64)` is an even cheaper approximation and there is some subset of valid inputs that would trigger it and for which we'd then need to compute the actual `max(x, y) + ceil(log2(min(x, y)))`.

So it basically just comes down to handling the 99% of inputs that can't possibly overflow as cheaply as possible. This is always fine because `lzcnt` is a native instruction for Arm, Arm64, and WASM. It is part of our default R2R target for x86/x64 and if you happen to be on x86/x64 that is pre-lzcnt (more than 11 years old at this point), you end up getting `(sizeof(T) * 8 - 1) ^ BSR` instead. For Arm32 and x86, you at worst get an extra branch and addition in here to account for the upper vs lower halves of the 64-bit input.

For the subset of cases that hit the "it could actually be under 64-bits", you then need to do `log2` which just uses the already computed `lzcnt` as input. So we remain with an incredibly efficient fast path that minimizes the total amount of work and doesn't need to do the extra multiplications required to check if bits `64-127` are non-zero.

jkotas replied:
> So it basically just comes down to handling the 99% of inputs that can't possibly overflow as cheaply as possible. 

Right. It is 99% inputs from distribution found in real-world programs. The current implementation is optimized to be as cheap as possible for multiplying two <32-bit numbers. It is quite likely it aligns with the 99% inputs distribution found in real-world programs.

jkotas replied:
This conversation made me check the perf of the rewrite. The rewrite clocks 1.25x slower on my quick lame micro-benchmark. Have you done https://github.com/dotnet/runtime/pull/98858#issuecomment-1986388593 for checked multiplication?

```
using System;
using System.Diagnostics;

for (;;)
{
    Stopwatch sw = new();
    sw.Start();
    Work(10_000_000_000);
    Console.WriteLine(sw.ElapsedMilliseconds.ToString());
}

static ulong Work(ulong iters)
{
    ulong sum = 0;
    for (ulong i = 0; i < iters; i += 123)
    {
        sum += checked(i * 54321);
    }
    return sum;
}
```

MichalPetryka replied:
> Have you done [#98858 (comment)](https://github.com/dotnet/runtime/pull/98858#issuecomment-1986388593) for checked multiplication?

I haven't but here the code should be unchanged so this difference is probably from RyuJIT generating worse codegen than MSVC on X86.

jkotas replied:
> Are you checking with Checked CoreLib btw? 

I always run perf tests on release builds of everything.

One of the problem seems to be that BigMul is not always inlined:

![image](https://github.com/dotnet/runtime/assets/6668460/d77a36ae-0a9e-4889-8ff1-8a092ff5ba4e)

It may be a good idea to split this PR into multiple smaller one so that each of the changes gets the right level of scrutiny.

jkotas replied:
The existing code (both the C/C++ version in regular CoreCLR and the managed native AOT version) does not use the BigMul helper, so it does not have this perf bug.

MichalPetryka replied:
> The existing code (both the C/C++ version in regular CoreCLR and the managed native AOT version) does not use the BigMul helper, so it does not have this perf bug.

Putting `AggressiveInlining` on `BigMul` like I did in #100027 makes the regression be about 2.9% on CoreCLR on my machine, I'd argue that's good enough in exchange for sharing the implementation between runtimes. It will also likely improve with future optimizations to the JIT so I'd assume it's gonna get better than native in the future.

> It may be a good idea to split this PR into multiple smaller one so that each of the changes gets the right level of scrutiny.

While I'd agree with this before, I feel like splitting the PR at this point after most review was done would only complicate things here.

jkotas replied:
> I feel like splitting the PR at this point after most review was done would only complicate things here.

It would allow me or other maintainers to get these changes merged with higher confidence and less work. In general, multiple smaller PRs are less total work for maintainers to get through than one big PR.

jkotas replied:
I have peeled off first chunk into https://github.com/dotnet/runtime/pull/100222

jkotas replied:
https://github.com/dotnet/runtime/pull/100222 is merged with number of adjustments. I think the next good-sized chunk to peel off into a separate PR is everything to do with the long multiplication. @MichalPetryka Are you interested in doing that?

MichalPetryka replied:
I'll look into splitting this up tomorrow.

jkotas commented:
/azp run runtime-extra-platforms, runtime-nativeaot outerloop

jkotas commented:
/azp run runtime-nativeaot-outerloop

MichalPetryka commented:
The `Build linux-arm64 Release NativeAOT_Checked_Libs` failure seems like it could be related, not sure why the Checked cast helper would not throw with `float -> int` anymore though and why it'd fail only there.

jkotas commented:
I expected that it would fail in some other combinations too. We have too many possible combinations in the test matrix and we are not able to test all of them.

MichalPetryka committed:
Call FMod directly

jkotas commented:
I have extracted the remaining good changes from this PR to https://github.com/dotnet/runtime/pull/102129
